tmail-backend/jmap/extensions-postgres/src/main/java/com/linagora/tmail/james/jmap/firebase/PostgresFirebaseSubscriptionDAO.java

```diff
@@ -0,0 +1,145 @@
+                .where(USER.eq(username.asString()))
+                .and(ID.in(idsMatching.stream().map(FirebaseSubscriptionId::value)
+                    .toList()))))
+            .map(this::toSubscription);
+
+        if (ids.size() <= IN_CLAUSE_MAX_SIZE) {
+            return queryPublisherFunction.apply(ids);
+        } else {
+            return Flux.fromIterable(Iterables.partition(ids, IN_CLAUSE_MAX_SIZE))
+                .flatMap(queryPublisherFunction);
+        }
+    }
+
+    public Mono<Void> deleteByUsername(Username username) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.deleteFrom(TABLE_NAME)
+            .where(USER.eq(username.asString()))));
+    }
+
+    public Mono<Void> deleteByUsernameAndId(Username username, FirebaseSubscriptionId id) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.deleteFrom(TABLE_NAME)
+            .where(USER.eq(username.asString()))
+            .and(ID.eq(id.value()))));
+    }
+
+    public Mono<Set<TypeName>> updateType(Username username, FirebaseSubscriptionId id, Set<TypeName> newTypes) {
+        Preconditions.checkNotNull(newTypes, "newTypes should not be null");
+        return postgresExecutor.executeRow(dslContext -> Mono.from(dslContext.update(TABLE_NAME)
+                .set(TYPES, newTypes.stream().map(TypeName::asString).toArray(String[]::new))
+                .where(USER.eq(username.asString()))
+                .and(ID.eq(id.value()))
+                .returning(TYPES)))
+            .map(this::extractTypes);
+    }
+
+    public Mono<ZonedDateTime> updateExpireTime(Username username, FirebaseSubscriptionId id, ZonedDateTime newExpire) {
+        Preconditions.checkNotNull(newExpire, "newExpire should not be null");
+        return postgresExecutor.executeRow(dslContext -> Mono.from(dslContext.update(TABLE_NAME)
+                .set(EXPIRES, newExpire.toOffsetDateTime())
+                .where(USER.eq(username.asString()))
+                .and(ID.eq(id.value()))
+                .returning(EXPIRES)))
+            .map(record -> OFFSET_DATE_TIME_ZONED_DATE_TIME_FUNCTION.apply(record.get(EXPIRES)));
+    }
+
+    private FirebaseSubscription toSubscription(Record record) {
+        return new FirebaseSubscription(new FirebaseSubscriptionId(record.get(ID)),
+            record.get(DEVICE_CLIENT_ID),
+            record.get(FCM_TOKEN),
+            toExpires(record),
+            CollectionConverters.asScala(extractTypes(record)).toSeq());
+    }
+
+    private FirebaseSubscriptionExpiredTime toExpires(Record record) {
+        return new FirebaseSubscriptionExpiredTime(OFFSET_DATE_TIME_ZONED_DATE_TIME_FUNCTION.apply(record.get(EXPIRES)));
+    }
+
+    private Set<TypeName> extractTypes(Record record) {
+        return Arrays.stream(record.get(TYPES))
+            .map(string -> typeStateFactory.parse(string).right().get())
```

Comment:
Do we handle "no longer existing types" stored on the DB? 
How?
What's the impact?
What's the alternative?

Reply:
> Do we handle "no longer existing types" stored on the DB?
How?

It is likely not the case as of today. (many place in our code as well)

> What's the impact?

Likely:
- FirebaseSubscription/get and PushSubscription/get would fail
- Pushing user notifications via Push gateway / FCM would fail because we can not deserialize the "no longer existing types".

> What's the alternative?

Be more lenient parsing the types I guess.

But is that a soon viable issue when we give up a supported type?

Reply:
Let's not fix that as off now but let's document the case in a ticket.

IMO being lenient + WARN log would be the bet when reading our data store.

Strict enforcement however on what is set by our clients.

Reply:
Can you open the ticket?

Reply:
> Can you open the ticket?

Yes. let me.

Reply:
https://github.com/linagora/tmail-backend/issues/958